spring-messaging/src/main/java/org/springframework/messaging/rsocket/DefaultRSocketRequester.java

```diff
@@ -310,12 +349,27 @@ private <T> Flux<T> retrieveFlux(ResolvableType elementType) {
 			if (isVoid(elementType)) {
 				return payloadFlux.thenMany(Flux.empty());
 			}
-
+			return payloadFlux.switchOnFirst((signal, innerFlux) -> {
+				Payload firstPayload = signal.get();
+				return firstPayload == null ? decodeFLux(innerFlux, elementType):
+						decodeFlux(firstPayload, innerFlux, elementType);
+			});
+		}
+		@SuppressWarnings("unchecked")
+		public <T> Flux<T> decodeFLux(Flux<Payload> payloads,ResolvableType elementType){
 			Decoder<?> decoder = strategies.decoder(elementType, dataMimeType);
-			return payloadFlux.map(this::retainDataAndReleasePayload).map(dataBuffer ->
+			return payloads.map(this::retainDataAndReleasePayload).map(dataBuffer ->
 					(T) decoder.decode(dataBuffer, elementType, dataMimeType, EMPTY_HINTS));
 		}
 
+		@SuppressWarnings("unchecked")
+		public <T> Flux<T> decodeFlux(Payload firstPayload, Flux<Payload> payloads,ResolvableType elementType){
+			MimeType contentMime =  messageContentType(firstPayload);
+			Decoder<?> decoder = strategies.decoder(elementType, contentMime);
+			return payloads.map(this::retainDataAndReleasePayload).map(dataBuffer ->
+					(T) decoder.decode(dataBuffer, elementType, contentMime, EMPTY_HINTS));
+		}
```

Comment:
Likewise on the requester side, we should be aware if a request specific content type is being used, and if not then avoid using the `switchOnFirst` operator.